crates/net/network/src/transactions.rs

```diff
@@ -1074,116 +1271,341 @@ struct Peer {
+    /// the channel to the peer's session task.
     ///
     /// This filters all announced hashes that are already in flight, and requests the missing,
     /// while marking the given peer as an alternative peer for the hashes that are already in
     /// flight.
     fn request_transactions_from_peer(
         &mut self,
-        mut announced_hashes: Vec<TxHash>,
+        new_announced_hashes: Vec<TxHash>,
         peer: &Peer,
-    ) -> bool {
+        metrics_increment_egress_peer_channel_full: impl FnOnce(),
+    ) -> Option<Vec<TxHash>> {
         let peer_id: PeerId = peer.request_tx.peer_id;
-        // 1. filter out inflight hashes, and register the peer as fallback for all inflight hashes
-        announced_hashes.retain(|&hash| {
-            match self.inflight_hash_to_fallback_peers.entry(hash) {
-                Entry::Vacant(entry) => {
-                    // the hash is not in inflight hashes, insert it and retain in the vector
-                    entry.insert(vec![peer_id]);
-                    true
-                }
-                Entry::Occupied(mut entry) => {
-                    // the hash is already in inflight, add this peer as a backup if not more than 3
-                    // backups already
-                    let backups = entry.get_mut();
-                    if backups.len() < MAX_ALTERNATIVE_PEERS_PER_TX {
-                        backups.push(peer_id);
-                    }
-                    false
-                }
-            }
-        });
 
-        // 2. request all missing from peer
-        if announced_hashes.is_empty() {
-            // nothing to request
-            return false
+        if self.active_peers.len() as u32 >= MAX_CONCURRENT_TX_REQUESTS {
+            debug!(target: "net::tx",
+                peer_id=format!("{peer_id:#}"),
+                hashes=format!("[{}]", new_announced_hashes.iter().fold(String::new(), |acc, &hash| acc + &format!("{hash:#},"))),
+                limit=MAX_CONCURRENT_TX_REQUESTS,
+                "limit for concurrent `GetPooledTransactions` requests reached, dropping request for hashes to peer"
+            );
+            return Some(new_announced_hashes)
+        }
+
+        let Some(inflight_count) = self.active_peers.get_or_insert(peer_id, || 0) else {
+            warn!(target: "net::tx",
+                peer_id=format!("{peer_id:#}"),
+                hashes=format!("[{}]", new_announced_hashes.iter().fold(String::new(), |acc, &hash| acc + &format!("{hash:#},"))),
+                "failed to cache active peer in schnellru::LruMap, dropping request to peer"
+            );
+            return Some(new_announced_hashes)
+        };
+
+        if *inflight_count >= MAX_CONCURRENT_TX_REQUESTS_PER_PEER {
+            debug!(target: "net::tx",
+                peer_id=format!("{peer_id:#}"),
+                hashes=format!("[{}]", new_announced_hashes.iter().fold(String::new(), |acc, &hash| acc + &format!("{hash:#},"))),
```

Comment:
I believe this can be done with itertools::format_with

Reply:
must have stashed that change, saw Dani's comment in an issue, thanks